src/components/plots/plot.ts

```diff
@@ -261,26 +276,52 @@ module Plottable {
      * This function makes sure that all of the scales in this._projections
      * have an extent that includes all the data that is projected onto them.
      */
-    protected _updateScaleExtents() {
-      d3.keys(this._projections).forEach((attr: string) => this._updateScaleExtent(attr));
-    }
-
-    public _updateScaleExtent(attr: string) {
-      var projector = this._projections[attr];
-      if (projector.scale) {
-        this._datasetKeysInOrder.forEach((key) => {
-          var plotDatasetKey = this._key2PlotDatasetKey.get(key);
-          var dataset = plotDatasetKey.dataset;
-          var plotMetadata = plotDatasetKey.plotMetadata;
-          var extent = dataset._getExtent(projector.accessor, projector.scale._typeCoercer, plotMetadata);
-          var scaleKey = this.getID().toString() + "_" + key;
-          if (extent.length === 0 || !this._isAnchored) {
-            projector.scale._removeExtent(scaleKey, attr);
-          } else {
-            projector.scale._updateExtent(scaleKey, attr, extent);
+    protected _updateAllExtents() {
+      d3.keys(this._projections).forEach((attr: string) => { this._updateExtentsForAttr(attr); });
+      var scales: Scale<any, any>[] = [];
+      d3.keys(this._projections).forEach((attr: string) => {
+        var scale = this._projections[attr].scale;
+        if (scale != null && scales.indexOf(scale) === -1) {
+          scales.push(scale);
+        }
+      });
```

Comment:
I should have commented here to start with.

Maybe...

Object.keys(this._projections).map((attr) => this._projectionrs[attr].scale).filter((scale) => scale != null).reduce((scale, uniqScales) => if (uniqScales.indexOf(scale) === -1) { return uniqScales.concat(scale); });

Lots and lots of pure transformations